src/Nethermind/Nethermind.AccountAbstraction/Source/UserOperationPool.cs

```diff
@@ -0,0 +1,338 @@
+            _userOperationSortedPool = userOperationSortedPool;
+            _userOperationSimulator = userOperationSimulator;
+                
+            // topic hash emitted by a successful user operation
+            _userOperationEventTopic = new Keccak("0xc27a60e61c14607957b41fa2dad696de47b2d80e390d0eaaf1514c0cd2034293");
+
+            MemoryAllowance.MemPoolSize = accountAbstractionConfig.UserOperationPoolSize;
+
+            _broadcaster = new UserOperationBroadcaster(logger);
+
+            _blockTree.BlockAddedToMain += OnBlockAdded;
+
+            ProcessNewBlocks();
+        }
+
+        private void OnBlockAdded(object? sender, BlockReplacementEventArgs e)
+        {
+            try
+            {
+                _headBlocksChannel.Writer.TryWrite(e);
+            }
+            catch (Exception exception)
+            {
+                if (_logger.IsError)
+                    _logger.Error(
+                        $"Couldn't correctly add or remove user operations from UserOperationPool after processing block {e.Block!.ToString(Block.Format.FullHashAndNumber)}.", exception);
+            }
+        }
+
+        private void ProcessNewBlocks()
+        {
+            Task.Factory.StartNew(async () =>
+            {
+                while (await _headBlocksChannel.Reader.WaitToReadAsync())
+                {
+                    while (_headBlocksChannel.Reader.TryRead(out BlockReplacementEventArgs? args))
+                    {
+                        try
+                        {
+                            ReAddReorganizedUserOperations(args.PreviousBlock);
+                            RemoveProcessedUserOperations(args.Block);
+                        }
+                        catch (Exception e)
+                        {
+                            if (_logger.IsDebug) _logger.Debug($"UserOperationPool failed to update after block {args.Block.ToString(Block.Format.FullHashAndNumber)} with exception {e}");
+                        }
+                    }
+                }
+            }, TaskCreationOptions.LongRunning).ContinueWith(t =>
+            {
+                if (t.IsFaulted)
+                {
+                    if (_logger.IsError) _logger.Error($"UserOperationPool update after block queue failed.", t.Exception);
+                }
+            });
+        }
+
+        private void ReAddReorganizedUserOperations(Block? previousBlock)
+        {
+            if (previousBlock is not null && _removedUserOperations.ContainsKey(previousBlock.Number))
+            {
+                foreach (UserOperation op in _removedUserOperations[previousBlock.Number])
+                {
+                    AddUserOperation(op);
+                }
+            }
+        }
+
+        public IEnumerable<UserOperation> GetUserOperations()
+        {
+            return _userOperationSortedPool.GetSnapshot();
+        }
+
+        public ResultWrapper<Keccak> AddUserOperation(UserOperation userOperation)
```

Comment:
a bit of leak of RPC type (ResultWrapper)

Reply:
how else can we return errors nicely?